packages/langium/src/workspace/profiler.ts

```diff
@@ -0,0 +1,189 @@
+        this.activeCategories = activeCategories ?? new Set<ProfilingCategory>(this.allCategories);
+    }
+
+    isActive(category: ProfilingCategory): boolean {
+        return this.activeCategories.has(category);
+    }
+
+    start(...categories: ProfilingCategory[]): void {
+        if (!categories) {
+            // Create a new set with all categories (immutable copy)
+            this.activeCategories = new Set(this.allCategories);
+        }
+        else {
+            categories.forEach(category => this.activeCategories.add(category));
+        }
+    }
+
+    stop(...categories: ProfilingCategory[]): void {
+        if (!categories) {
+            this.activeCategories.clear();
+        } else {
+            categories.forEach(category => this.activeCategories.delete(category));
+        }
+    }
+    createTask(category: ProfilingCategory, taskId: string): ProfilingTask {
+        if (!this.isActive(category)) {
+            throw new Error(`Category "${category}" is not active.`);
+        }
+        console.log(`Creating profiling task for '${category}.${taskId}'.`);
+        return new ProfilingTask((record: ProfilingRecord) => this.records.add(category, this.dumpRecord(category, record)), taskId);
+    }
+
+    private dumpRecord(category: string, record: ProfilingRecord): ProfilingRecord {
+        console.info(`Task ${category}.${record.identifier} executed in ${record.duration.toFixed(2)}ms and ended at ${record.date.toISOString()}`);
+
+        const result: Array<{ name: string, count: number, duration: number }> = [];
+        for (const key of record.entries.keys()) {
+            const values = record.entries.get(key);
+            const duration = values.reduce((p, c) => p + c);
+            result.push({ name: `${record.identifier}.${key}`, count: values.length, duration: duration });
+        }
+
+        // sum all duration
+        const taskInternalDuration = record.duration - result.map(r => r.duration).reduce((a, b) => a + b, 0);
+
+        result.push({ name: record.identifier, count: 1, duration: taskInternalDuration });
+
+        result.sort((a, b) => b.duration - a.duration);
+        function Round(value: number) { return Math.round(100 * value) / 100; }
+        console.table(result.map(e => { return { Element: e.name, Count: e.count, 'Self %': Round(100 * e.duration / record.duration), 'Time (ms)': Round(e.duration) }; }));
+        return record;
+    }
+    getRecords(...categories: ProfilingCategory[]): Stream<ProfilingRecord> {
+        if (categories.length === 0) {
+            // return all records
+            return this.records.values();
+        } else {
+            // return records for the given categories
+            return this.records.entries().filter((e) => categories.some(c => c === e[0])).flatMap(e => e[1]);
+        }
+    }
+    protected readonly records: MultiMap<string, ProfilingRecord> = new MultiMap();
+}
+
+interface ProfilingRecord {
```

Comment:
Minor: Since this type is actually part of public API (see `ProfilingTask.constructor`), we should export this.
```suggestion
export interface ProfilingRecord {
```